### ISSUE TITLE
TextInput foreground_color property ignored in version 1.9.1

### ISSUE THREAD

**Issue by trushworth** [kivy/kivy#4056]
Setting or changing the foreground_color property on a TextInput widget does not seem to work for kivy 1.9.1.  Digging into the code shows that the _get_line_options() function constructs a dictionary of options for the Label class with things like font_size and font_name, but does not include color.  The attached one-line diff fixes that, but I'm not sure it will do everything needed when the ListProperty is changed since I set the color once at the beginning of my app and never change it.

[textinput_py_diff.txt](https://github.com/kivy/kivy/files/157897/textinput_py_diff.txt)

**Comment by dessant** [comment]
The linked diff:
```diff
--- textinput.py-orig	2016-03-03 15:42:22.000000000 -0800
+++ textinput.py	2016-03-03 15:43:48.000000000 -0800
@@ -2066,6 +2066,7 @@
             self._line_options = kw = {
                 'font_size': self.font_size,
                 'font_name': self.font_name,
+                'color': self.foreground_color,
                 'anchor_x': 'left',
                 'anchor_y': 'top',
                 'padding_x': 0,
```

This should also be looked into:
https://github.com/kivy/kivy/blob/1dd9bcd4e8d7c715bb287636c826d3174fb142ed/kivy/data/style.kv#L179

**Comment by trushworth** [comment]
I'm a very long way from any kind of expert, but I've looked at it and the last instruction in the default style for the TextInput.canvas.before is to set the correct foreground color depending on the state of the input (disabled, hint, etc.).  The very first thing that Label does is to set canvas Color to a constant, which would seem to make the last instruction in the TextInput.canvas.before pointless.  Where all of this is likely to break down is if you are using markup, and the text is being wrapped over multiple lines, and the markup is setting the color too, then the Label setting the color is going to mess up the markup.

To make things a little less broken, how about using the line of code from the TextInput style directly in the line options dictionary instead of the simple self.foreground_color I used in the diff above?  Since each Label sets the color, the last color setting in the TextInput.canvas.before could be dropped too.  It won't fix any issues with markup, but non-markup TextInputs should be good.

**Comment by trushworth** [comment]
Oops. the older browser I was using isn't supported well enough to attach a revised diff.  Here it is from a different browser:

[textinput_py_diff.txt](https://github.com/kivy/kivy/files/159254/textinput_py_diff.txt)

**Comment by dessant** [comment]
@trushworth, the best is if you paste the diff inline, in code tags, ;)

**Comment by trushworth** [comment]
re: diffs -  OK, might as well start with the last one then:
```diff
--- textinput.py-orig   2016-03-03 15:42:22.000000000 -0800
+++ textinput.py  2016-03-04 11:37:33.000000000 -0800
@@ -2066,6 +2066,7 @@
             self._line_options = kw = {
                 'font_size': self.font_size,
                 'font_name': self.font_name,
+                'color': self.disabled_foreground_color if self.disabled else (self.hint_text_color if not self.text else self.foreground_color),
                 'anchor_x': 'left',
                 'anchor_y': 'top',
                 'padding_x': 0,
```
I was just thinking about fixing the color for multi-line text with markup.  I don't know if the evaluation order would allow the following, could someone who understands it better let me know?  I am using color here as the attribute I'm interested in, but this mechanism would apply to all the markup controlled attributes.  What would be nice is if we could set the color for the first Label using the code from the second diff, but then extract the final color (in effect at end of the Label text, assuming markup text records the current color somewhere) of each Label and use the extracted color as the starting color of the next Label.  It might even be possible to put a link to the preceding Label in the Label itself and to use the link when not None to determine the starting color.

Thoughts anyone?

**Comment by dessant** [comment]
@trushworth could you share a runnable example which does not work? I can help with a fix.

I have tested foreground_color and it does seem to work with kivy master.

**Comment by trushworth** [comment]
I'll try to come up with one, but I've run out of free time, it may be a
week or two before I can get back to kivy.  We want an example with a
single TextInput widget, not the mess I'm working on.

On 16-03-06 11:04 , dessant wrote:
> @trushworth could you share a runnable example which does not work? I can help with a fix.
> 
> I have tested foreground_color and it does seem to work with kivy master.
> 
[snip]

Regards,

-- 
Tom Rushworth

**Comment by dessant** [comment]
@trushworth, just fill this in, it should not take more than a couple of minutes if you have already reproduced it locally.

```python
from kivy.app import App
from kivy.lang import Builder
from kivy.factory import Factory
from kivy.uix.boxlayout import BoxLayout

kv = """
<Test>:
    TextInput:
    
"""

Builder.load_string(kv)


class Test(BoxLayout):
    pass

class TestApp(App):
    def build(self):
        return Test()

if __name__ == '__main__':
    TestApp().run()
```

**Comment by trushworth** [comment]
Hi,

OK, I think I need to explain further.  I am learning kivy by
implementing a simple hierarchical ToDo application.  Essentially a
TreeView in a ScrollView, with a TreeViewNode containing a small icon on
the left and a TextInput on the right.

A TextInput seems to set the text color for all its lines by setting a
color as the last instruction in self.canvas.before and simply relying
on that color to be the current/most recent one when the time comes to
draw the text.

I want to combine the TreeViewNode open/close icon with status
(done/not-done) info from the To Do list, so I need different imagery
than the default TreeViewNode imagery.  The TreeViewNode creates its
default open/close "chevron" imagery as defined in data/style.kv (approx
line 220) in the canvas.before group.  To use different imagery I need
to clear the canvas.before and supply my own drawing code.  This all
works, I get the images I want in the colors I want and everything
changes as it should (so far).  Except for the text :).

All of the examples I've seen, including the one in the docstring for
TreeView shows the TreeViewNode class as the last class in the super
class list of any custom node class.  For my ToDo_Node class I declared:

```python
class ToDo_Node(TextInput,TreeViewNode):
```

This means that the TreeViewNode's __init__() function will get called
after the TextInput's __init__() function.  If the TreeViewNode's
__init__() function happens to set any colors in self.canvas.before then
TextInput's assumption about the current color will be wrong.  The
default style for TreeViewNode in data/style.kv sets a color in the
canvas.before if it gets used, but even if it doesn't, my custom imagery
drawing also sets a color there, so the Labels in the TextInput end up
with the wrong color for the text in any case.

The fact that TextInput is relying on some color being the last color
set in canvas.before seems like an unexpected and bug-prone coupling
that makes TextInput tricky to use as a super class.  It's not a bug as
such, more like an open invitation to create a bug :).

I've fixed my current issue by carefully creating the color I want for
the text as the very last thing in canvas.before in my own ToDo_Node's
__init__() function, after calling the super class inits and clearing
canvas.before and drawing my own imagery.  I'd like to see a more robust
mechanism for setting the text color in a TextInput widget, but I don't
need it at this point and this issue could be closed if you folks feel
it appropriate.

However, the reason I brought up the issue of markup is that where I
eventually want to go with this ToDo app is to have markup text in the
node, for example I will eventually want to be able to have something like:

```
The quick [color=#ff0000ff]red fox jumped\nover the lazy [/color]brown dog.
```

and end up with two lines showing, where the last part of line 1 and the
first part of line 2 are in red, not black.  My simple "fix" for a more
robust text color setting would result in all of line 2 being black, so
the "fix" is not good enough for the markup text scenario.

Experimenting with the code fragment @dessant provided shows that
TextInput doesn't do markup text at present and thinking about it makes
it clear there will have to be quite a bit of work done (by myself if
no-one else wants it or gets to it before me) before it can.  (And BTW,
I do not mean this as a criticism or a request that someone else do
something, kivy is ambitious and very cool as it is today.  I was led by
how capable it is now to hope for that little bit more :), so keep up
the good work!)


On 16-03-06 13:29 , dessant wrote:
> @trushworth, just fill this in, it should not take more than
 a couple of minutes if you have already reproduced it locally.
> 
> ```python
> from kivy.app import App
> from kivy.lang import Builder
> from kivy.factory import Factory
> from kivy.uix.boxlayout import BoxLayout
> 
> kv = """
> <Test>:
>     TextInput:
>     
> """
> 
> Builder.load_string(kv)
> 
> 
> class Test(BoxLayout):
>     pass
> 
> class TestApp(App):
>     def build(self):
>         return Test()
> 
> if __name__ == '__main__':
>     TestApp().run()
> ```
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/kivy/kivy/issues/4056#issuecomment-193000841
> 
Regards,

-- 
Tom Rushworth
